manager/volume.go

```diff
@@ -613,6 +616,46 @@ func (m *VolumeManager) Expand(volumeName string, size int64) (v *longhorn.Volum
 		return nil, fmt.Errorf("cannot expand volume before replica scheduling success")
 	}
 
+	kubernetesStatus := &v.Status.KubernetesStatus
+	if kubernetesStatus.PVCName != "" && kubernetesStatus.LastPVCRefAt == "" {
+		pvc, err := m.ds.GetPersistentVolumeClaim(kubernetesStatus.Namespace, kubernetesStatus.PVCName)
+		if err != nil {
+			return nil, err
+		}
+
+		resourceSize := resource.MustParse(strconv.FormatInt(size, 10))
+
+		// TODO: Should check for pvc.Spec.Resources.Requests.Storage() here, once upgrade API to v0.18.x.
+		storageSpecValue := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+		// TODO: Should check for pvc.Status.CapacityStorage() here, once upgrade API to v0.18.x.
+		storageStatusValue := pvc.Status.Capacity[corev1.ResourceStorage]
+
+		if storageSpecValue.Cmp(resourceSize) > 0 {
+			return nil, fmt.Errorf("cannot expand persistent volume claim %v with current size %v to a smaller size %v", v.Name, v.Spec.Size, size)
+		}
+
+		if storageSpecValue.Cmp(resourceSize) == 0 && storageStatusValue.Cmp(resourceSize) == 0 && v.Spec.Size == size {
+			logrus.Infof("persistent volume claim %v expansion is not necessary since current size %v == %v", v.Name, v.Spec.Size, size)
+			return v, nil
+		}
+
+		if storageSpecValue.Cmp(resourceSize) < 0 {
+			pvc.Spec.Resources = corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: resourceSize,
+				},
+			}
+
+			logrus.Infof("Persistent Volume Claim %v expansion from %v to %v requested", v.Name, v.Spec.Size, size)
+			_, err = m.ds.UpdatePersistentVolumeClaim(kubernetesStatus.Namespace, pvc)
+			if err != nil {
+				return nil, err
+			}
+
+			return v, nil
+		}
```

Comment:
Please just do two conditions to avoid missing conditions to flow to the original logic.

if storageSpecValue.Cmp(resourceSize) < 0 {
... 
} else {
... // the current volume size is large or eqaul to the requrested size.
}

Reply:
changed